code/datums/client_image_controller.dm

```diff
@@ -0,0 +1,359 @@
+		Removes a mob/mind's ability to see images in "cimg_key" group
+		Validation is stackable, so they'll be still capable of seeing stuff when validation is given multiple times.
+		Both(mob/mind) does the same thing, but mob/mind difference.
+
+
+		< What's best to use between mob and mind? >
+			GLOB.cimg_controller.validate_mind("heretic_pierces", heretic_mind)
+		You want heretics have this always even if their body is changed (likely brain transplant)
+		aghost will still be capable of seeing stuff, because their mind is validated and ghost has that mind.
+
+			GLOB.cimg_controller.validate_mob("heretic_pierces", target_mob)
+		If a curator wears goggles that can see "heretic_pierces", we want to use this proc instead of mind.
+		aghost will rip off the images because they become a ghost, but their aghost is not validated.
+		once aghost goes back to their mob, they'll see the stuff again.
+*/
+
+/datum/cimg_controller
+	/// list of /datum/cimg_holder datum types
+	var/list/cimg_holders = list()
+
+	/// a quick key list for on mob login
+	var/list/cimgkey_by_mob = list()
+	/// a quick key list for on mob login (mob login checks mind too)
+	var/list/cimgkey_by_mind = list()
+
+/datum/cimg_holder
+	/// list of images for a group
+	var/list/bound_images = list()
+	/// identical to 'bound_images' but we do this |= instead of += because it can exist in multiple groups
+	var/list/shared_bound_images = list()
+
+	/// mobs who can see that
+	var/list/valid_mobs = list()
+	/// minds who can see that
+	var/list/valid_minds = list()
+
+/// adds an image to a key group. You must call `cut_client_images()` for qdel.
+/// * [is_shared_image]: Set this TRUE if an image is in multiple holders.
+/datum/cimg_controller/proc/stack_client_images(cimg_key, client_images, is_shared_image = FALSE)
+	var/datum/cimg_holder/cimg_holder = cimg_holders[cimg_key]
+	if(!cimg_holder)
+		cimg_holder = new
+		cimg_holders[cimg_key] = cimg_holder
+
+	if(is_shared_image)
+		cimg_holder.shared_bound_images += client_images // list works
+	else
+		cimg_holder.bound_images += client_images // list works
+	cimg_holder._realize_to_validated(client_images, is_shared_image = is_shared_image)
+
+/// removes added image from a key group.
+/// * [is_shared_image]: Set this TRUE if an image is in multiple holders.
+/datum/cimg_controller/proc/cut_client_images(cimg_key, client_images, is_shared_image = FALSE)
+	var/datum/cimg_holder/cimg_holder = cimg_holders[cimg_key]
+	if(!cimg_holder)
+		return
+
+	if(is_shared_image)
+		cimg_holder.shared_bound_images -= client_images // list works
+	else
+		cimg_holder.bound_images -= client_images // list works
+	cimg_holder._disappear_from_validated(client_images)
+
+/// radical version of cut_client_images() - delete a group
+/// usually done when an image group is dedicated for a short duration or personal.
+/// Avoid using this unless a client image group is NOT common
+/datum/cimg_controller/proc/nuke_client_image_holder(cimg_key)
```

Comment:
This should be named more appropriately, if it deletes an entire group then name it cut_group or something. Nuke client image holder makes it sound like its a last result for when the system is broken.

Also why shouldn't it be used if the client image group is common? This should be able to take in any group and remove it from everyone that has it, even if every player on the server has it. (For the cultist effect where it makes all non-cultists see everyone else as a cultist, when it finishes we want to mass-remove all cultist images from everyone that has it)

Reply:
This is now renamed to `nullify_client_vision_holder()`

This exists when a client image group is temporary, very private, or something.
For example, `"hallucination_[FAST_REF(src)]"` from hallucination will be only temporary for the moment when a mob has that, and we will want "nuke" this group once this hallucination is over.
This nuke removes the key from everyone even if their key wasn't given from the hallucination (i.e. when Ghost can see their observed one's hallucination? This should be stripped.)